File: scripts/emoji-enhancement/local-matte.py
"""JSON-lines worker. Install rembg[cpu]==2.0.84 in an external Python 3.12 venv."""
import json
import sys
from pathlib import Path
from rembg import new_session, remove

session = new_session('birefnet-general-lite', providers=['CPUExecutionProvider'])
for line in sys.stdin:
    job = json.loads(line)
    try:
        output = Path(job['output'])
        output.parent.mkdir(parents=True, exist_ok=True)
        output.write_bytes(remove(Path(job['input']).read_bytes(), session=session))
        print(json.dumps({'id': job['id'], 'ok': True}), flush=True)
    except Exception as error:
        print(json.dumps({'id': job['id'], 'error': str(error)}), flush=True)

File: scripts/emoji-enhancement/local-matte.mjs
import {spawn} from 'node:child_process';import fs from 'node:fs/promises';
let child,pending=new Map(),nextId=0,buffer='';
export async function localMatte(input,output){
 try{await fs.access(output);return {model:'birefnet-general-lite',cached:true}}catch{}
 if(!child){child=spawn(process.env.EMOJI_MATTE_PYTHON??'/tmp/emoji-enhancement-venv/bin/python',['scripts/emoji-enhancement/local-matte.py'],{env:{...process.env,OMP_NUM_THREADS:'4'},stdio:['pipe','pipe','pipe']});child.stdout.on('data',chunk=>{buffer+=chunk;while(buffer.includes('\n')){const split=buffer.indexOf('\n'),line=buffer.slice(0,split);buffer=buffer.slice(split+1);try{const r=JSON.parse(line),p=pending.get(r.id);if(p){pending.delete(r.id);r.ok?p.resolve({model:'birefnet-general-lite'}):p.reject(Error(r.error))}}catch{}}});child.on('error',error=>{for(const p of pending.values())p.reject(error);pending.clear()});child.on('exit',code=>{for(const p of pending.values())p.reject(Error('Local matte worker exited '+code));pending.clear();child=undefined});child.stderr.on('data',()=>{});}
 const id=nextId++;return new Promise((resolve,reject)=>{pending.set(id,{resolve,reject});child.stdin.write(JSON.stringify({id,input,output})+'\n')});
}
export function closeLocalMatte(){child?.stdin.end()}
